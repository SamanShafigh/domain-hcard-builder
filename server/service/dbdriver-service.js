var MongoClient = require('mongodb').MongoClient;

async function connect({ dbUri, dbName }) {
  return new Promise((resolve, reject) => {
    MongoClient.connect(dbUri, {useNewUrlParser: true}, (err, client) => {
      if (err) {
        return reject(err)
      }
      var db = client.db(dbName)
      resolve(db)
    });
  })
}

module.exports = {
  connect
}