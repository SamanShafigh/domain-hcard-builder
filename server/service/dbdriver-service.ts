import { MongoClient } from 'mongodb';

export async function dbConnect({ dbUri, dbName }: any) {
  return new Promise((resolve: any, reject: any) => {
    MongoClient.connect(dbUri, {useNewUrlParser: true}, (err: any, client: any) => {
      if (err) {
        return reject(err)
      }
      var db = client.db(dbName)
      resolve(db)
    });
  })
}