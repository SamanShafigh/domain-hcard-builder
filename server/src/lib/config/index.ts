import HCardT from '../../type';

export function makeConfig(env: HCardT.Env): HCardT.Config {
  var config = <HCardT.Config>{
    staticRoot: env.STATIC_ROOT || './client/dist',
    serverPort: env.PORT || 3030,
    renderMode: env.RENDER_MODE || 'ssr',
    dbUri: env.DB_URL || 'mongodb://localhost:27017',
    dbName: env.DB_NAME || 'domain-hcard',
  }

  return config;
}