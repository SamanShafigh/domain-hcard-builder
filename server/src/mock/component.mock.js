module.exports = {
  default: 'h1'
}