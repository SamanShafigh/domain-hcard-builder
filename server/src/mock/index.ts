import HCardT from './type';

export const mockEnv = <HCardT.Env>{
  STATIC_ROOT: 'root',
  PORT: 1234,
  RENDER_MODE: 'ssr',
  DB_URL: 'mongodb://localhost',
  DB_NAME: 'test'
};

export const mockConfig = {
  staticRoot: './client/dist',
  ssrPath: './dist/main.js',
  spaPath: './dist/_index.html',
  serverPort: 3030,
  renderMode: 'ssr',
  dbUri: 'mongodb://localhost',
  dbName: 'hcard',
}

export const mockUser = {
  givenName: 'givenName',
  surname: 'surname',
  email: 'email',
  phone: 'phone',
  houseNumber: 'houseNumber',
  street: 'street',
  suburb: 'suburb',
  state: 'state',
  postcode: 'postcode',
  country: 'country'
}

export const mockInvalidUser = {
  givenName: 'givenName',
  email: 'email',
  houseNumber: 'houseNumber',
  suburb: 'suburb',
  postcode: 'postcode',
  invalidData: 'some Invalid data'
}

export const mockCtx = {
  meta: {
    userId: '1234',
    noscript: true,
  }
}

export const mockCtxWithBody = {
  meta: {
    userId: '1234',
    noscript: true,
  },
  request: {
    body: mockUser
  }
}

export const mockDbDriver = {
  findOne: (model, query, options) => (Promise.resolve({ data: mockUser })),
  save: (model, query, data, options) => ({ model, query, data, options }),
  update: (model, query, data, options) => ({ model, query, data, options }),
}

export const mockMongoDbDriver = {
  collection: (model) => ({
    findOne: (query, options) => Promise.resolve({ model, query, options }),
    update: (query, data, options) => Promise.resolve({ model, query, data, options}),
  })
}

export const mockReactComponentPath = '../mock/component.mock.js'
export const mockHtmlPagePath = '../mock/page.mock.html'