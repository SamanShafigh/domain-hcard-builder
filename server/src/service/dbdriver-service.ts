import { MongoClient } from 'mongodb';
import HCardT from '../type';

/**
 * This is a default DB storage driver currently we support
 */
class MongoDbDriver implements HCardT.DbDriver {
  db: any
  constructor(db: any) {
    this.db = db;
  }

  /**
   * Find an entity
   * @param model 
   * @param query 
   * @param options 
   */
  async findOne(model: string, query: any, options?: any) {
    return await this.db
      .collection(model)
      .findOne(query, options);
  }

  /**
   * Save an entity
   * @param model 
   * @param query 
   * @param data 
   * @param options 
   */
  async save(model: string,query: any, data: any, options?: any) {
    await this.db
      .collection(model)
      .update(
        query, 
        {$set: data}, 
        Object.assign({upsert: true}, options)
      );
  }

  /**
   * Update an entity
   * @param model 
   * @param query 
   * @param data 
   * @param options 
   */
  async update(model: string, query: any, data: any, options?: any) {
    await this.db
      .collection(model)
      .update(
        query, 
        {$set: data}, 
        Object.assign({upsert: true}, options)
      );
  }  
}

/**
 * Make a Data storage driver, the default one is Mongo, but I can make it 
 * more advanced to return different type of drivers
 * 
 * @param dbUri 
 * @param dbName 
 */
export async function makeDbDriver(dbUri: string, dbName: string): Promise<any> {
  return new Promise((resolve: any, reject: any) => {
    // Just make a default mongoDB driver
    MongoClient.connect(dbUri, {useNewUrlParser: true}, (err: any, client: any) => {
      if (err) {
        return reject(err)
      }
      var db = client.db(dbName)
      var dbDriver = new MongoDbDriver(db);

      resolve(dbDriver)
    });
  })
}