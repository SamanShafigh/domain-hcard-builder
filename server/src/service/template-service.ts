import { createReadStream } from 'fs';
import * as React from 'react';
import * as ReactDom from 'react-dom/server';

interface Global { React: any; }
declare var global: Global;
global.React = React;

function html(body: string) {
  return `
  <!DOCTYPE html>
  <html lang="en">
    <head>
      <meta charset="utf-8">
      <meta http-equiv="X-UA-Compatible" content="IE=edge">
      <meta name="viewport" content="width=device-width, initial-scale=1">
      <title>Live hCard Preview</title>
  
      <link href="css/bootstrap.min.css" rel="stylesheet" >
      <link href="css/main.css" rel="stylesheet">
    </head>
  
    <body>
      ${body}
    </body>
  </html>
  `;
} 

export function render(path: string, props: any) {
  const component = require(path).default;
  var body = ReactDom.renderToString(
    React.createElement(component, props)
  );

  return html(body);
}

export function serve(path: string) {
  return createReadStream(`${__dirname}/${path}`)
}
