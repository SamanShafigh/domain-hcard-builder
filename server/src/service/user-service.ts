import HCardT from '../type';
import makeUser from '../lib/model/user-model';

const MODEL_NAME = 'user'

/**
 * UserService is to abstract the main logic related to User
 */
class UserService implements HCardT.UserService {
  dbDriver: HCardT.DbDriver
  
  constructor(dbDriver: HCardT.DbDriver) {
    this.dbDriver = dbDriver;
  }

  /**
   * Get a user and if user does not exist returns null
   * @param userId 
   */
  async getUser(userId: string) {
    var result = await this.dbDriver
      .findOne(MODEL_NAME, { userId }, { _id: 0 });
    
    return result? makeUser(result.data) : null;
  }  

  /**
   * Submit a user data
   * @param userId 
   * @param data 
   */
  async submitUser(userId: string, data: any) {
    const user = makeUser(data);
    return await this.dbDriver
      .update(MODEL_NAME, { userId }, { data: user });
  }

  /**
   * Update a user parameter
   * @param userId 
   * @param key 
   * @param value 
   */
  async updateUser(userId: string, key: string, value: string) {
    var reference = `data.${key}`;

    return await this.dbDriver
      .update(MODEL_NAME, { userId }, { [reference]: value });
  }
}

export default UserService;
