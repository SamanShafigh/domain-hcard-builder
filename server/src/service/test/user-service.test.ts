import UserService from '../user-service';
import * as dbdriver from '../../service/dbdriver-service';
import { mockUser, mockDbDriver } from '../../lib/mock';
import HCardT from '../type';

let userService: HCardT.UserService;
jest.mock('../../service/dbdriver-service');

beforeAll(async () => {
  // Mocking my DB driver
  dbdriver.makeDbDriver.mockImplementationOnce(() => mockDbDriver)

  // Make my UserService
  const mockDriver = dbdriver.makeDbDriver();
  userService = new UserService(mockDriver)
});

describe('Test user service', () => {
  test('test to get a user', async () => {
    var result = await userService.getUser(1234);
    expect(result).toMatchSnapshot();
  });

  test('test to save a user', async () => {
    var result = await userService.submitUser(1234, mockUser);
    expect(result).toMatchSnapshot();
  });

  test('test to update a user', async () => {
    var result = await userService.updateUser(1234, 'givenName', 'saman');
    expect(result).toMatchSnapshot();
  });

  test('test to get a null user', async () => {
    // Make my UserService with a driver that returns null
    dbdriver.makeDbDriver.mockImplementationOnce(() => ({
      findOne: (model, query, options) => (Promise.resolve(null)),
    }))
    var userService = new UserService(dbdriver.makeDbDriver())

    var result = await userService.getUser(1234);
    expect(result).toEqual(null);
  });

  test('test to make sure we not exposing critical information', async () => {
    // Make my UserService with a driver that returns some unexpected data
    dbdriver.makeDbDriver.mockImplementationOnce(() => ({
      findOne: (model, query, options) => (
        Promise.resolve({ data: {_id: 'we should not return this'}})
      ),
    }))
    var userService = new UserService(dbdriver.makeDbDriver())

    try {
      await userService.getUser(1234)
      expect(true).toBeFalsy();
    } catch (e) {
      expect(e).toBeInstanceOf(Error);
      expect(e.message).toBe('Param _id is not defined in User model')
    }
  });
});