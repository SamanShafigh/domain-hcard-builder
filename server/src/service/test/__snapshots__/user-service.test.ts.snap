// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test user service test to get a user 1`] = `
User {
  "country": "country",
  "email": "email",
  "givenName": "givenName",
  "houseNumber": "houseNumber",
  "phone": "phone",
  "postcode": "postcode",
  "state": "state",
  "street": "street",
  "suburb": "suburb",
  "surname": "surname",
}
`;

exports[`Test user service test to save a user 1`] = `
Object {
  "data": Object {
    "data": User {
      "country": "country",
      "email": "email",
      "givenName": "givenName",
      "houseNumber": "houseNumber",
      "phone": "phone",
      "postcode": "postcode",
      "state": "state",
      "street": "street",
      "suburb": "suburb",
      "surname": "surname",
    },
  },
  "model": "user",
  "options": undefined,
  "query": Object {
    "userId": 1234,
  },
}
`;

exports[`Test user service test to update a user 1`] = `
Object {
  "data": Object {
    "data.givenName": "saman",
  },
  "model": "user",
  "options": undefined,
  "query": Object {
    "userId": 1234,
  },
}
`;
